File: 2021/day06.py
def children_per_fish(state, days, sols={}):
    if days <= state:
        return 0
    if (state, days) not in sols:
        sols[(state, days)] = 1 + (days-state-1)//7 + sum(children_per_fish(8, days-d) for d in range(state+1, days, 7))
    return sols[(state, days)]


def part1(fish_states):
    print(sum(1 + children_per_fish(state, days=80) for state in fish_states))


def part2(fish_states):
    print(sum(1 + children_per_fish(state, days=256) for state in fish_states))


with open("input/day06.txt") as f:
    v = list(map(int, f.readline().split(',')))
    part1(v)
    part2(v)

File: 2021/day06.go
package main

import (
	"aoc/utils"
	"fmt"
	"strings"
)

func shift(states []int, days int) int {
	for ; days > 0; days-- {
		_0s := states[0]
		for i := 0; i < 8; i++ {
			states[i] = states[i+1]
		}
		states[8] = _0s
		states[6] += states[8]
	}
	return utils.Sum(states)
}

// Part1 of day6
func Part1(states []int) {
	fmt.Println(shift(states, 80))
}

// Part2 of day6
func Part2(states []int) {
	fmt.Println(shift(states, 256))
}

func main() {
	states := make([]int, 9)
	for _, s := range utils.ToIntSliceOrDie(strings.Split(
		utils.ReadFileContentOrDie("./input/day06.txt"), ",")) {
		states[s]++
	}
	Part1(utils.Copy(states))
	Part2(states)
}
